Update deps, add additional unit tests, add signals-polyfill and signia, update results

transitive-bullshit opened the pull request:
This PR rebases on top of #12 by @NullVoxPopuli. It updates all deps and contains the changes I needed to get everything working.

- updates all deps to latest (except for `compostate` since when I updated to latest it no longer worked)
- adds [TC39 Signals Proposal](https://github.com/tc39/proposal-signals) [polyfill](https://github.com/proposal-signals/signal-polyfill) framework via #12 
- adds [signia](https://github.com/tldraw/signia) framework
- adds some additional unit tests
- changes `mobx` and `@vue/reactivity` to use shallow signals (to make sure we're comparing apples to apples)
- adds some additional unit tests to ensure all frameworks are working as expected
- fixes several bugs and inconsistencies with the existing framework implementations
- adds a prettier config which matches the existing formatting
- updates the readme results ([preview here](https://github.com/transitive-bullshit/js-reactivity-benchmark/tree/feature/update))

For some reason several frameworks are hanging or throwing errors on the `cellx` benchmark, so I've disabled it for now. The same goes for the largest dynamic benchmark, which I've commented out for now.

NullVoxPopuli commented:
Looks like we have some iterating to do on the performance of the signal polyfill.

milomg commented:
Wow these new results are really exciting!

milomg commented:
Quick update: I'm aiming to test and merge this weekend

robbiespeed commented:
I don't think this should merge until the callcount assertions are uncommented out, and cellx bench is re-enabled.

Was there a reason dynamic bench was reduced from 5 to 1 runs? If some FWs are failing 5 runs of some dynamic tests then that could be hinting at a memory leak in those FWs.

If there's a certain benchmark that is commonly failing for some frameworks then maybe a case can be added to the configs like `testCellX`.

transitive-bullshit commented:
> I don't think this should merge until the callcount assertions are uncommented out, and cellx bench is re-enabled.
> 
> Was there a reason dynamic bench was reduced from 5 to 1 runs? If some FWs are failing 5 runs of some dynamic tests then that could be hinting at a memory leak in those FWs.
> 
> If there's a certain benchmark that is commonly failing for some frameworks then maybe a case can be added to the configs like `testCellX`.

Great feedback & catches. I had disabled certain things while debugging different frameworks and not everything got put back in the toy bin :) Will update shortly.

milomg commented:
My current changes:

* Remove `  (globalThis as any).__DEV__ = true;`
* Remove funding.yml
* Add back in cellx but turned off 2500 and 5000 for now

transitive-bullshit commented:
> Was there a reason dynamic bench was reduced from 5 to 1 runs? If some FWs are failing 5 runs of some dynamic tests then that could be hinting at a memory leak in those FWs.

I think the reason I changed this is because some of these frameworks take 10+ seconds for 1 run on the dynamic tests, so it's a minute per test per framework which is pretty annoying to wait for. It's not nearly as bad for the better optimized frameworks, but TC39 signals polyfill and angular signals are a pain to benchmark, making the tweak => iteration cycle difficult. Maybe in a follow-up PR, we can address how long it takes to run a benchmark end-to-end?

> I don't think this should merge until the callcount assertions are uncommented out, and cellx bench is re-enabled.

Done.

> My current changes in addition to the commented out checks:
> 
> * Remove `  (globalThis as any).__DEV__ = true;`
> * Remove funding.yml
> * Add back in cellx but turned off 2500 and 5000 for now

Great catches; done.

See [`db852c5` (#13)](https://github.com/milomg/js-reactivity-benchmark/pull/13/commits/db852c55b93f5eb6718e27cd8592486aa0cea58d) for details. There was still one kairo assertion which was failing for multiple frameworks (on the `unstable` test), so I added a comment there.

transitive-bullshit commented:
After making these changes, here's the latest output from `pnpm bench`: [out.csv](https://github.com/user-attachments/files/17268787/out.csv)

transitive-bullshit commented on src/dynamicBench.ts:
```diff
@@ -10,7 +10,7 @@ import { fastestTest } from "./util/benchRepeat";
 export async function dynamicBench(
   frameworkTest: FrameworkInfo,
-  testRepeats = 1
+  testRepeats = 5
```
changing it back to the original value

milomg commented:
Yeah, for now I was just replacing the CSV file [here](https://public.tableau.com/app/profile/modder.me/viz/ReactivitySplit/PerformanceResults_1) with an updated one, but it should be possible to create a chartjs/d3 script to autogenerate results

milomg commented on src/frameworks/solid.ts:
```diff
@@ -22,7 +22,7 @@ export const solidFramework: ReactiveFramework = {
     };
   },
-  effect: (fn) => createEffect(fn),
+  effect: (fn) => createRenderEffect(fn),
```
Why this change?

transitive-bullshit replied:
same reason as above; the new unit tests weren't passing without it.

milomg commented on src/util/dependencyGraph.ts:
```diff
@@ -54,30 +49,60 @@ export function runGraph(
 
-  for (let i = 0; i < iterations; i++) {
+    sum = readLeaves.reduce((total, leaf) => leaf.read() + total, 0);
+  } else {
```
This allows for an algorithm that defers all changes until after a batch is finished to only walk the graph once instead of `iterations` times (especially because we have no verification on what leaf.read returns.

Would it be possible to have all frameworks use the same path? (and in a future change I'll add assertions to ensure that every framework's leaf.read is actually getting the right result)

milomg commented on src/frameworks/reactively.ts:
```diff
@@ -16,7 +16,10 @@ export const reactivelyFramework: ReactiveFramework = {
   },
-  effect: (fn) => new Reactive(fn, true),
+  effect: (fn) => {
+    fn();
```
Why is this needed? Eagerly evaluating effects should probably be done by calling the required function in the framework (stabilize) instead of calling the effect function twice (which gives reactively and oby a disadvantage if the effect is slow because they can't memoize it)

transitive-bullshit replied:
the new unit tests that I added to ensure we're comparing apples to apples across the different frameworks only work with this added.

happy to change it, but I feel like it's important for the unit tests across all frameworks to work the same way.

milomg replied:
I agree with having tests to ensure that the benchmark is fair, although I think the unit tests may be more strict than we expect. Afaik we don't have any tests that read or write during the build stage (they always return functions that are run outside of the build stage).

Then, making withBuild run a stabilize tick after construction should get the effects to run as intended

milomg replied:
This seems to pass tests for me

```ts
export const reactivelyFramework: ReactiveFramework = {
  name: "@reactively",
  signal: (initialValue) => {
    const r = new Reactive(initialValue);
    return {
      write: (v) => r.set(v),
      read: () => r.get(),
    };
  },
  computed: (fn) => {
    const r = new Reactive(fn);
    return {
      read: () => r.get(),
    };
  },
  effect: (fn) => {
    return new Reactive(fn, true);
  },
  withBatch: (fn) => {
    fn();
    stabilize();
  },
  withBuild: (fn) => {
    const out = fn();
    stabilize();
    return out;
  },
};
```

milomg commented:
I've been trying to incorporate features from this PR into the main branch. Is there anything else that's missing for you (if so I'd love to add it), or can I close this PR?
